step-grid-parent/step-grid-proxy/src/main/java/step/grid/proxy/GridProxy.java

```diff
@@ -0,0 +1,219 @@
+        readTimeout = gridProxyConfiguration.getGridReadTimeout();
+    }
+
+    //Allow override for junits
+    protected void overrideRestClient(Client client) {
+        this.client = client;
+    }
+
+    public void handleRegistrationMessage(RegistrationMessage message) throws MalformedURLException {
+        String agentUrl = message.getAgentRef().getAgentUrl();
+        // replace by proxyfied url (proxy base url + context root) and maintain the mapping
+        message.getAgentRef().setAgentUrl(gridProxyUrl + "/" + getContextRoot(agentUrl));
+        //Forward registration message to grid server
+        try (Response r = client.target(gridUrl + "/grid/register").request().property(ClientProperties.READ_TIMEOUT, readTimeout)
+                    .property(ClientProperties.CONNECT_TIMEOUT, connectTimeout).post(Entity.entity(message, MediaType.APPLICATION_JSON))) {
+
+            r.readEntity(String.class);
+        } catch (ProcessingException e) {
+            if(e.getCause() instanceof java.net.ConnectException) {
+                logger.error("Unable to reach " + gridUrl + " while proxyfying agent registration from " + agentUrl + " (java.net.ConnectException: "+e.getCause().getMessage()+")");
+            } else {
+                logger.error("Error while proxyfying registration request to " + gridUrl + " from " + agentUrl, e);
+            }
+            //Rethrow exception as response to the Agent
+            throw e;
+        }
+    }
+
+    /**
+     * This method return the registered context root for the provided agent URL or generates a unique one for new registration
+     * <p>Note: parsing to URL might cause performance issue, could be replaced by URL encoding otherwise</p>
+     * <p>In case of malformed URL the exceptions is rethrown</p>
+     *
+     * @param agentUrl the agent Url used to generate the context root
+     * @return the unique context root for this agent
+     */
+    private String getContextRoot(String agentUrl) throws MalformedURLException {
+        //Not using computeIfAbsent for rethrowing MalformedURLException
+        if (agentUrlToContextRoot.containsKey(agentUrl)) {
+            return agentUrlToContextRoot.get(agentUrl);
+        } else {
+            URL url = new URL(agentUrl);
+            String host = url.getHost();
+            int port = url.getPort();
+            String contextRoot = host + "-" + port;
+            agentUrlToContextRoot.put(agentUrl, contextRoot);
+            contextRootToAgentUrl.put(contextRoot, agentUrl);
+            logger.info("Registering new agent from url '" + agentUrl + "' with contextRoot: " + contextRoot);
+            return contextRoot;
+        }
+    }
+
+    public Response forwardGetFileRequest(String fileId, String version) throws IOException {
+        try (Response fromGrid = client.target(gridUrl +  "/grid/file/" + fileId + "/" + version).request().property(ClientProperties.READ_TIMEOUT, readTimeout)
```

Comment:
@jeromecomte we currently have the same implementation on the agent side, but it seems the value is overriden in the configuration provided in the distribution by 
gridReadTimeout: 20000
Should I change the default value in the java class to 20 seconds ?

Reply:
@david-stephan We can keep the default value in the code low but should ensure that the value of the distribution is at always higher that the one defined in the agent to avoid timeouts in the proxy before they occur in the agent

Reply:
Implementation chagned to have indivial timeouts configuration, distrbution updated to have values higher than the agent / grid-server